File: python/binaryTreeDiameter.py
class BinaryTree:
    def __init__(self, value):
        self.value = value
        
        self.left = None
        self.right = None
        
    
    def GetDiameter(self):
        if self == None:
            return TreeInfo(0, 0)
        
        leftTreeInfo = self.left.GetDiameter()
        rightTreeInfo = self.right.GetDiameter()
        
        maxdiameter = max(leftTreeInfo.diameter, rightTreeInfo.diameter, (rightTreeInfo.height + leftTreeInfo.height))
        height = max(leftTreeInfo.height, rightTreeInfo.height) + 1
        
        return TreeInfo(maxdiameter, height)
    
        
class TreeInfo:
    def __init__(self, diameter, height):
        self.diameter = diameter
        self.height = height
        
        
def getMaxDiameter(tree):
    return tree.GetDiameter().diameter

File: go/binaryTreeDiameter/binaryTreeDiameter.go
package main


type BinaryTree struct {
	Value int

	Left *BinaryTree
	Right *BinaryTree
}


type TreeInfo struct {
	Diameter int
	Height int
}


// Time -> 0(N)
func (tree *BinaryTree) MaxDiameter() *TreeInfo {
	if tree == nil {return &TreeInfo{Diameter: 0, Height: 0}}

	leftTreeInfo := tree.Left.MaxDiameter()
	rightTreeInfo := tree.Right.MaxDiameter()


	maxdiameter := max((leftTreeInfo.Height + rightTreeInfo.Height), leftTreeInfo.Diameter, rightTreeInfo.Diameter)
	height := max(leftTreeInfo.Height, rightTreeInfo.Height) + 1

	return &TreeInfo{Diameter: maxdiameter, Height: height}
}


func GetDiameter(tree *BinaryTree) int {
	return tree.MaxDiameter().Diameter
}
